src/Raven.Server/Documents/AI/Embeddings/QueryEmbeddingsCacher.cs

```diff
@@ -0,0 +1,113 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Raven.Server.Background;
+using Raven.Server.Documents.TransactionMerger.Commands;
+using Raven.Server.ServerWide.Context;
+using Sparrow.Server;
+
+namespace Raven.Server.Documents.AI.Embeddings;
+
+public class QueryEmbeddingsCacher : BackgroundWorkBase
+{
+    private readonly DocumentDatabase _database;
+
+    private readonly ConcurrentQueue<EmbeddingGenerationItem> _embeddingsQueue;
+    private readonly AsyncManualResetEvent _mre;
+    private int _approximateCount;
+    public QueryEmbeddingsCacher(DocumentDatabase database, CancellationToken shutdown) : base(database.Name, database.Loggers.GetLogger<QueryEmbeddingsCacher>(), shutdown)
+    {
+        _database = database;
+        _embeddingsQueue = new ConcurrentQueue<EmbeddingGenerationItem>();
+        _mre = new AsyncManualResetEvent();
+    }
+
+    protected override async Task DoWork()
+    {
+        while (CancellationToken.IsCancellationRequested == false)
+        {
+            await _mre.WaitAsync(CancellationToken);
+
+            _mre.Reset();
+
+            var mightBeMore = CacheEnqueuedEmbeddings();
+            
+            if (mightBeMore)
+                _mre.Set();
+        }
+    }
+
+    internal bool CacheEnqueuedEmbeddings()
+    {
+        var mightBeMore = false;
+        
+        var maxBatchSize = _database.Configuration.Ai.QueryEmbeddingsGenerationMaxCacheBatchSize;
+
+        var payload = new List<EmbeddingGenerationItem>(Math.Min(_approximateCount, maxBatchSize));
+
+        while (_embeddingsQueue.TryDequeue(out var item))
```

Comment:
wouldnt BlockingCollection be more suitable here?

Reply:
The idea is to enqueue multiple items and _then_ signal the cacher to start processing.  `BlockingCollection` would be signaled immediately that there is something to take so that could result in smaller batches.